src/contracts/ODSaviour.sol

```diff
@@ -1,12 +1,168 @@
+ */
+contract ODSaviour is AccessControl, IODSaviour {
+  using Math for uint256;
+  using Assertions for address;
 
-  function enableVault(uint256 _tokenId) external {}
+  // solhint-disable-next-line modifier-name-mixedcase
+  bytes32 public constant SAVIOUR_TREASURY = keccak256(abi.encode('SAVIOUR_TREASURY'));
+  bytes32 public constant PROTOCOL = keccak256(abi.encode('PROTOCOL'));
+
+  uint256 public liquidatorReward;
+
+  address public saviourTreasury;
+  address public protocolGovernor;
+  address public liquidationEngine;
+
+  IVault721 public vault721;
+  IOracleRelayer public oracleRelayer;
+  IODSafeManager public safeManager;
+  ISAFEEngine public safeEngine;
+  ICollateralJoinFactory public collateralJoinFactory;
+
+  mapping(uint256 _vaultId => bool _enabled) private _enabledVaults;
+  mapping(bytes32 _cType => IERC20 _tokenAddress) private _saviourTokenAddresses;
+
+  /**
+   * @param _init The SaviourInit struct;
+   */
+  constructor(SaviourInit memory _init) {
+    saviourTreasury = _init.saviourTreasury.assertNonNull();
+    protocolGovernor = _init.protocolGovernor.assertNonNull();
+    vault721 = IVault721(_init.vault721.assertNonNull());
+    oracleRelayer = IOracleRelayer(_init.oracleRelayer.assertNonNull());
+    safeManager = IODSafeManager(address(vault721.safeManager()));
+    liquidationEngine = ODSafeManager(address(safeManager)).liquidationEngine(); // todo update @opendollar package to include `liquidationEngine` - PR #693
+    collateralJoinFactory = ICollateralJoinFactory(_init.collateralJoinFactory.assertNonNull());
+    safeEngine = ISAFEEngine(address(safeManager.safeEngine()));
+    liquidatorReward = _init.liquidatorReward;
+
+    if (_init.saviourTokens.length != _init.cTypes.length) revert LengthMismatch();
+
+    // solhint-disable-next-line  defi-wonderland/non-state-vars-leading-underscore
+    for (uint256 i; i < _init.cTypes.length; i++) {
+      _saviourTokenAddresses[_init.cTypes[i]] = IERC20(_init.saviourTokens[i].assertNonNull());
+    }
+    _setupRole(SAVIOUR_TREASURY, saviourTreasury);
+    _setupRole(PROTOCOL, protocolGovernor);
+    _setupRole(PROTOCOL, liquidationEngine);
+  }
+
+  function isEnabled(uint256 _vaultId) external view returns (bool _enabled) {
+    _enabled = _enabledVaults[_vaultId];
+  }
+
+  function addCType(bytes32 _cType, address _tokenAddress) external onlyRole(SAVIOUR_TREASURY) {
+    _saviourTokenAddresses[_cType] = IERC20(_tokenAddress);
+    emit CollateralTypeAdded(_cType, _tokenAddress);
+  }
+
+  function cType(bytes32 _cType) public view returns (address _tokenAddress) {
+    return address(_saviourTokenAddresses[_cType]);
+  }
+
+  function setLiquidatorReward(uint256 _newReward) external onlyRole(PROTOCOL) {
+    liquidatorReward = _newReward;
+    emit LiquidatorRewardSet(_newReward);
+  }
+
+  /**
+   * @dev
+   */
+  function setVaultStatus(uint256 _vaultId, bool _enabled) external onlyRole(SAVIOUR_TREASURY) {
+    _enabledVaults[_vaultId] = _enabled;
+
```

Comment:
Add validation that ctype is enabled for this vault type

Reply:
done in saviour branch.